pkg/mutation/path/parser/node.go

```diff
@@ -66,6 +67,118 @@ func (r Path) String() string {
 	return result.String()
 }
 
+func (r Path) GetValueAtLocation(object interface{}) ([]string, error) {
```

Comment:
I think this code can go away if we retrieve the value mid-mutation

Reply:
Given the below requirement that we should batch the inputs for each provider, I'd need this function to gather all the values before we initiate the requests if they aren't filled by previous, non-external data mutations with the same path.

If we were to retrieve the value mid-mutation, we can only call a request to the provider with that particular. Since mutators work independently, I am not sure if there is a way to coordinate batching+parallelizing.

Reply:
So the idea would be:

* call the [mutate](https://github.com/open-policy-agent/gatekeeper/blob/11223ec9adf1228a1e2d458eb31d07810819f8ff/pkg/mutation/mutators/core/mutation_function.go#L72-L95) function, which may write out placeholders if external data mutators are present
* Take the output of the mutate function (an `unstructured` resource) and recurse through that resource gathering any placeholder objects that exist
* Collate the placeholder objects to dedupe, batch, etc.
* Send out requests
* recurse through the object again, replacing placeholder objects with concrete values

WRT getting the original value, pre-mutation... this should not be necessary.

Let's consider field `A`, with value "orig".

It has 4 mutators, applied in the following order:

1. Set value to "one"
2. Call remote provider "R2"
3. Set value to "three"
4. Call remote provider "R4"
5. Call remote provider "R5"

The value of the field at every step (with 0 being the original value) is:

0. "orig"
1. "one"
2. `Placeholder{provider: "R2", value: "one"}`
3. "three"
4. `Placeholder{provider: "R4", value: "three"}`
5. `Placeholder{provider: "R5", value: "three"}`

Notice that placeholders can be clobbered, either by a constant value or a new promise to call a provider. In this way, irrelevant mutators don't result in an outbound request.

Lets take a more complex example, here are two mutators in the specified call order:

1. Add a sidecar `{name: "my_logger", image: "logging:latest"}`
2. Replace `spec.containers[name:*].image` with a call to the provider "ConvertTagToHash" 

Using the same format as above (0 is the original object):

0.

```yaml
spec:
  containers:
  - name: "frontend"
    image: "my-app:v1"
```

1.

```yaml
spec:
  containers:
  - name: "frontend"
    image: "my-app:v1"
  - name: "my_logger"
    image: "logging:latest"
```

2.

```yaml
spec:
  containers:
  - name: "frontend"
    image: `Placeholder{Provider: "ConvertTagToHash", Value: "my-app:v1"}`
  - name: "my_logger"
    image:  `Placeholder{Provider: "ConvertTagToHash", Value: ""logging:latest"}`
```

Note that the provider mutator must appear last here, lest it be clobbered by the sidecar mutator.

Traversing the mutated object, you'd find two fields of type `Placeholder{}`:

```golang
{Placeholder{Provider: "ConvertTagToHash", Value: "my-app:v1"}, Placeholder{Provider: "ConvertTagToHash", Value: ""logging:latest"}}
```

You would send out a call to ConvertTagToHash for keys: `["my-app:v1", "logging:latest"]`, lets say you get back:

```json
{
  "my-app:v1": "my-app:sha123",
  "logging:latest": "logging:sha456"
}
````

You would take that result, recurse over the mutated object, replacing placeholders with appropriate values, with the result:

```yaml
spec:
  containers:
  - name: "frontend"
    image: "my-app:sha123"
  - name: "my_logger"
    image: "logging:sha456"
```

If your concern is "what if there is no non-external data mutator?" The answer is that that shouldn't matter. Mutators are all calling the same `mutate()` function, with different logic for `Setter()`... setter would be assigning the Placeholder{} object.

Reply:
One other question... if we are retrieving values prior to starting mutation, how would we be able to handle the sidecar injection case?

Reply:
One interesting feature of this flow, demonstrated in the "mutation of field `A`" part of the megacomment:

R5 is being called against the output of mutator 3, not mutator 4. We could try to adjust that behavior, but that would lead to some providers needing to be called in serial, which has latency implications (see my max latency of ~1s for the mutation webhook above), so hopefully this is a reasonable tradeoff between functional behavior and strict truth.

Then again, both are premised on the idea that intentional chaining of external data mutators on a single field will be rare, so...

Reply:
Got it, if I understanad it correctl,y we would inject placeholder struct to the podspec even though placeholder is not part of the pod API, promising that we would resolve them to the appropriate types before we return the mutated object.

> One other question... if we are retrieving values prior to starting mutation, how would we be able to handle the sidecar injection case?

My thought was to perform all non-external data mutations first before retrieving the values for the external data providers. With your explanation above, I don't think it's necessary to retrieve the values anymore.

> R5 is being called against the output of mutator 3, not mutator 4.

That is the approach we are using when I discussed this issue with @sozercan. The source of truth for the mutator ordering would be in `orderedMutators` in the mutation system.

Reply:
> That is the approach we are using when I discussed this issue with @sozercan. The source of truth for the mutator ordering would be in orderedMutators in the mutation system.

Which is the approach? That the SOT is the value from Mutator 3?

In this case the thought isn't so much about the ordering (the example just presumes a certain order to clarify edge cases), but more "is ignoring mutator 4 the right thing?" I think both approaches are defensible, though def. worth seeing if we get user feedback once we have a prototype before committing to a final approach.

Reply:
> Which is the approach? That the SOT is the value from Mutator 3?

Yes. When creating R5 placeholder, we will use R4 placeholder's value at location, and then replace R4 placeholder with R5 placeholder.